### PR TITLE
ABDM: Log the errors instead of sending them as a response

### PR DESCRIPTION
**Pull request by ProCode2**
## Proposed Changes

- Added log for Exception with stack trace.
- replace Exception object with error string to be sent to client.

### Associated Issue

- #1815 

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [x] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

**Review by khavinshankar — APPROVED** [review]

**Review by Ashesh3 — APPROVED** [review]

**Review by rithviknishad — APPROVED** [review]

**Comment by khavinshankar** [comment]
@ProCode2 can you handle this issue (https://github.com/coronasafe/care_fe/pull/7000#issuecomment-1895353628) also in this pr.

To replicate the issue, In the front end,
1. Go to a patient consultation dashboard
2. Link an Abha number
3. Go to another patient's consultation dashboard
4. Link the same Abha number again

That will trigger the error you see in the comment mentioned. 

Have a conditional check if the Abha number already exists before creating the Abha number, and return a response with a neat error message.

File: care/abdm/api/viewsets/auth.py [hunk 1]
```diff
@@ -346,10 +356,14 @@ def post(self, request, *args, **kwargs):
...
         except Exception as e:
+            logger.warning(
+                f"Error: RequestDataView::post failed to notify (health-information/notify). Reason: {e}",
+                exc_info=True,
+            )
             return Response(
                 {
                     "detail": "Failed to notify (health-information/notify)",
-                    "error": str(e),
+                    "error": "Error: Failed to notify",
```
Comment (sainak) [inline care/abdm/api/viewsets/auth.py:366]:
```suggestion
```
no need for this key as we only use the detail to show the error message 
https://github.com/coronasafe/care_fe/blob/51b02c26666877f2b0b481f0476a9b34acba6743/src/Redux/fireRequest.tsx#L158

File: care/abdm/api/viewsets/healthid.py [hunk 2]
```diff
@@ -363,8 +366,15 @@ def link_via_qr(self, request):
                     }
                 )
             except Exception as e:
+                logger.warning(
+                    f"Error: ABDMHealthIDViewSet::link_via_qr failed to fetch nodes. Reason: {e}",
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:370]:
```suggestion
                    f"Error: ABDMHealthIDViewSet::link_via_qr failed to fetch modes. Reason: {e}",
```
typo

File: care/abdm/api/viewsets/healthid.py [hunk 3]
```diff
@@ -363,8 +366,15 @@ def link_via_qr(self, request):
...
+                    exc_info=True,
+                )
                 return Response(
-                    {"detail": "Failed to fetch modes", "error": str(e)},
+                    {
+                        "detail": "Failed to fetch modes",
+                        "error": "Error: ABDM Gateway node fetch failed",
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:376]:
```suggestion
```

File: care/abdm/api/viewsets/healthid.py [hunk 4]
```diff
@@ -419,8 +429,16 @@ def get_new_linking_token(self, request):
                 }
             )
         except Exception as e:
+            logger.warning(
+                f"Error: ABDMHealthIDViewSet::get_new_linking_token failed to fetch nodes. Reason: {e}",
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:433]:
```suggestion
                f"Error: ABDMHealthIDViewSet::get_new_linking_token failed to fetch modes. Reason: {e}",
```

File: care/abdm/api/viewsets/healthid.py [hunk 5]
```diff
@@ -419,8 +429,16 @@ def get_new_linking_token(self, request):
...
+                exc_info=True,
+            )
+
             return Response(
-                {"detail": "Failed to fetch modes", "error": str(e)},
+                {
+                    "detail": "Failed to fetch modes",
+                    "error": "Error: failed to fetch nodes.",
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:440]:
```suggestion
```

File: care/abdm/api/viewsets/healthid.py [hunk 6]
```diff
@@ -463,8 +481,16 @@ def add_care_context(self, request, *args, **kwargs):
...
+            logger.warning(
+                f"Error: ABDMHealthIDViewSet::add_care_context failed. Reason: {e}",
+                exc_info=True,
+            )
+
             return Response(
-                {"detail": "Failed to add care context", "error": str(e)},
+                {
+                    "detail": "Failed to add care context",
+                    "error": "Error: Failed to integrate care.",
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:492]:
```suggestion
```

File: care/abdm/api/viewsets/healthid.py [hunk 7]
```diff
@@ -496,8 +522,13 @@ def patient_sms_notify(self, request, *args, **kwargs):
...
             )
         except Exception as e:
+            logger.warning(
+                f"Error: ABDMHealthIDViewSet::patient_sms_notify failed to send SMS. Reason: {e}",
+                exc_info=True,
+            )
+
             return Response(
-                {"detail": "Failed to send SMS", "error": str(e)},
+                {"detail": "Failed to send SMS", "error": "Error: SMS not sent."},
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:531]:
```suggestion
                {"detail": "Failed to send SMS"},
```

File: data/medibase.json [hunk 8]
```diff
@@ -1144305,4 +1144305,4 @@
   "cims_class_link": "/india/drug/search?q=drugs+for+erectile+dysfunction&mtype=brand&code=10d",
   "cims_class": "Drugs for Erectile Dysfunction",
   "atc_classification": "G04BE03 - sildenafil ; Belongs to the class of drugs used in erectile dysfunction."
-}]
\ No newline at end of file
+}]
```
Comment (sainak) [inline data/medibase.json:1144308]: Revert this change

**Review by sainak — CHANGES_REQUESTED** [review]

**Comment by ProCode2** [comment]
> @ProCode2 can you handle this issue ([coronasafe/care_fe#7000 (comment)](https://github.com/coronasafe/care_fe/pull/7000#issuecomment-1895353628)) also in this pr.
> 
> To replicate the issue, In the front end,
> 
> 1. Go to a patient consultation dashboard
> 2. Link an Abha number
> 3. Go to another patient's consultation dashboard
> 4. Link the same Abha number again
> 
> That will trigger the error you see in the comment mentioned.
> 
> Have a conditional check if the Abha number already exists before creating the Abha number, and return a response with a neat error message.


@khavinshankar 
 by testing on https://care.ohc.network I traced it down to this function thats returning the error:

This creates a ABHA object if not exists and then adds ABHA details to patient.

Since `create_abha` is only creating the abha object if it does not exist it should be the `add_abha_details_to_patient` thats throwing error? So we need to not add abha details to a patient if that abha details is already linked with another patient?

Do I understand this correctly?
```python
def confirm_with_mobile_otp(self, request):
        data = request.data

        if ratelimit(request, "confirm_with_mobile_otp", [data["txnId"]]):
            raise CaptchaRequiredException(
                detail={"status": 429, "detail": "Too Many Requests Provide Captcha"},
                code=status.HTTP_429_TOO_MANY_REQUESTS,
            )

        serializer = VerifyOtpRequestPayloadSerializer(data=data)
        serializer.is_valid(raise_exception=True)
        response = HealthIdGateway().confirm_with_mobile_otp(data)
        abha_profile = HealthIdGateway().get_profile(response)

        # have a serializer to verify data of abha_profile
        abha_object = self.create_abha(
            abha_profile,
            {
                "access_token": response["token"],
                "refresh_token": response["refreshToken"],
                "txn_id": data["txnId"],
            },
        )

        if "patientId" in data:
            patient_id = data.pop("patientId")
            allowed_patients = get_patient_queryset(request.user)
            patient_obj = allowed_patients.filter(external_id=patient_id).first()
            if not patient_obj:
                raise ValidationError({"patient": "Not Found"})

            if not self.add_abha_details_to_patient(
                abha_object,
                patient_obj,
            ):
                return Response(
                    {"message": "Failed to add abha Number to the patient"},
                    status=status.HTTP_400_BAD_REQUEST,
                )

        return Response(
            {"id": abha_object.external_id, "abha_profile": abha_profile},
            status=status.HTTP_200_OK,
        )
```

File: care/abdm/api/viewsets/auth.py [hunk 9]
```diff
@@ -25,7 +28,14 @@ def post(self, request, *args, **kwargs):
         try:
             AbdmGateway().init(data["resp"]["requestId"])
         except Exception as e:
-            return Response({"error": str(e)}, status=status.HTTP_400_BAD_REQUEST)
+            logger.warning(
+                f"Error: OnFetchView::post failed while initialising ABDM Gateway, Reason: {e}",
+                exc_info=True,
+            )
+            return Response(
+                {"error": "Error: Initialising ABDM Gateway failed."},
```
Comment (sainak) [inline care/abdm/api/viewsets/auth.py:36]:
```suggestion
                {"detail": "Error: Initialising ABDM Gateway failed."},
```

File: data/medibase.json (hunk 8 above)
Comment (sainak) [inline data/medibase.json:1144308]: run `git restore --source=master -- data/medibase.json && git commit -anm "undo changes"` to revert this change

File: care/abdm/api/viewsets/healthid.py [hunk 10]
```diff
@@ -463,8 +479,16 @@ def add_care_context(self, request, *args, **kwargs):
...
+            logger.warning(
+                f"Error: ABDMHealthIDViewSet::add_care_context failed. Reason: {e}",
+                exc_info=True,
+            )
+
             return Response(
-                {"detail": "Failed to add care context", "error": str(e)},
+                {
+                    "detail": "Failed to add care context",
+                    "error": "Error: Failed to integrate care.",
```
Comment (sainak) [inline care/abdm/api/viewsets/healthid.py:490]:
```suggestion
```

**Review by sainak — CHANGES_REQUESTED** [review]

File: data/medibase.json (hunk 8 above)
Comment (ProCode2) [reply data/medibase.json]: Thanks done. Updated the other requested changes as well.

**Review by sainak — APPROVED** [review]

**Review by rithviknishad — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]
